File: dkplayer-players/exo/src/main/java/com/dueeeke/videoplayer/exo/ExoMediaSourceHelper.java
package com.dueeeke.videoplayer.exo;

import android.content.Context;
import android.net.Uri;
import android.text.TextUtils;

import com.google.android.exoplayer2.C;
import com.google.android.exoplayer2.ext.rtmp.RtmpDataSourceFactory;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.source.ProgressiveMediaSource;
import com.google.android.exoplayer2.source.dash.DashMediaSource;
import com.google.android.exoplayer2.source.hls.HlsMediaSource;
import com.google.android.exoplayer2.source.smoothstreaming.SsMediaSource;
import com.google.android.exoplayer2.upstream.DataSource;
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory;
import com.google.android.exoplayer2.upstream.DefaultHttpDataSource;
import com.google.android.exoplayer2.upstream.DefaultHttpDataSourceFactory;
import com.google.android.exoplayer2.upstream.HttpDataSource;
import com.google.android.exoplayer2.util.Util;

import java.lang.reflect.Field;
import java.util.Map;

public final class ExoMediaSourceHelper {

    private final String mUserAgent;
    private DataSource.Factory mDataSourceFactory;
    private Context mAppContext;
    private HttpDataSource.Factory mHttpDataSourceFactory;

    public ExoMediaSourceHelper(Context context) {
        mAppContext = context.getApplicationContext();
        mUserAgent = Util.getUserAgent(mAppContext, mAppContext.getApplicationInfo().name);
    }

    public MediaSource getMediaSource(String uri) {
        return getMediaSource(uri, null);
    }

    public MediaSource getMediaSource(String uri, Map<String, String> headers) {
        Uri contentUri = Uri.parse(uri);
        if ("rtmp".equals(contentUri.getScheme())) {
            return new ProgressiveMediaSource.Factory(new RtmpDataSourceFactory(null))
                    .createMediaSource(contentUri);
        }
        int contentType = inferContentType(uri);
        if (mDataSourceFactory == null) {
            mDataSourceFactory = getDataSourceFactory();
            if (mHttpDataSourceFactory != null) {
                setHeaders(headers);
            }
        }
        switch (contentType) {
            case C.TYPE_DASH:
                return new DashMediaSource.Factory(mDataSourceFactory).createMediaSource(contentUri);
            case C.TYPE_SS:
                return new SsMediaSource.Factory(mDataSourceFactory).createMediaSource(contentUri);
            case C.TYPE_HLS:
                return new HlsMediaSource.Factory(mDataSourceFactory).createMediaSource(contentUri);
            default:
            case C.TYPE_OTHER:
                return new ProgressiveMediaSource.Factory(mDataSourceFactory).createMediaSource(contentUri);
        }
    }

    public int inferContentType(String fileName) {
        fileName = Util.toLowerInvariant(fileName);
        if (fileName.contains(".mpd")) {
            return C.TYPE_DASH;
        } else if (fileName.contains(".m3u8")) {
            return C.TYPE_HLS;
        } else if (fileName.matches(".*\\.ism(l)?(/manifest(\\(.+\\))?)?")) {
            return C.TYPE_SS;
        } else {
            return C.TYPE_OTHER;
        }
    }

    /**
     * Returns a new DataSource factory.
     *
     * @return A new DataSource factory.
     */
    public DataSource.Factory getDataSourceFactory() {
        return new DefaultDataSourceFactory(mAppContext, getHttpDataSourceFactory());
    }

    /**
     * Returns a new HttpDataSource factory.
     *
     * @return A new HttpDataSource factory.
     */
    public DataSource.Factory getHttpDataSourceFactory() {
        if (mHttpDataSourceFactory == null) {
            mHttpDataSourceFactory = new DefaultHttpDataSourceFactory(
                    mUserAgent,
                    null,
                    DefaultHttpDataSource.DEFAULT_CONNECT_TIMEOUT_MILLIS,
                    DefaultHttpDataSource.DEFAULT_READ_TIMEOUT_MILLIS,
                    //http->https重定向支持
                    true);
        }
        return mHttpDataSourceFactory;
    }

    private void setHeaders(Map<String, String> headers) {
        if (headers != null && headers.size() > 0) {
            for (Map.Entry<String, String> header : headers.entrySet()) {
                String key = header.getKey();
                String value = header.getValue();
                //如果发现用户通过header传递了UA，则强行将HttpDataSourceFactory里面的userAgent字段替换成用户的
                if (TextUtils.equals(key, "User-Agent")) {
                    if (!TextUtils.isEmpty(value)) {
                        try {
                            Field userAgentField = mHttpDataSourceFactory.getClass().getDeclaredField("userAgent");
                            userAgentField.setAccessible(true);
                            userAgentField.set(mHttpDataSourceFactory, value);
                        } catch (Exception e) {
                            //ignore
                        }
                    }
                } else {
                    mHttpDataSourceFactory.getDefaultRequestProperties().set(key, value);
                }
            }
        }
    }

    public void setDataSourceFactory(DataSource.Factory factory) {
        mDataSourceFactory = factory;
    }

    public void setHttpDataSourceFactory(HttpDataSource.Factory factory) {
        mHttpDataSourceFactory = factory;
    }
}

File: dkplayer-players/exo/src/main/java/xyz/doikki/videoplayer/exo/ExoMediaSourceHelper.kt
package xyz.doikki.videoplayer.exo

import android.content.Context
import android.net.Uri
import android.text.TextUtils
import com.google.android.exoplayer2.C
import com.google.android.exoplayer2.MediaItem
import com.google.android.exoplayer2.database.StandaloneDatabaseProvider
import com.google.android.exoplayer2.ext.rtmp.RtmpDataSource
import com.google.android.exoplayer2.source.MediaSource
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import com.google.android.exoplayer2.source.dash.DashMediaSource
import com.google.android.exoplayer2.source.hls.HlsMediaSource
import com.google.android.exoplayer2.source.rtsp.RtspMediaSource
import com.google.android.exoplayer2.upstream.DataSource
import com.google.android.exoplayer2.upstream.DefaultDataSource
import com.google.android.exoplayer2.upstream.DefaultHttpDataSource
import com.google.android.exoplayer2.upstream.HttpDataSource
import com.google.android.exoplayer2.upstream.cache.Cache
import com.google.android.exoplayer2.upstream.cache.CacheDataSource
import com.google.android.exoplayer2.upstream.cache.LeastRecentlyUsedCacheEvictor
import com.google.android.exoplayer2.upstream.cache.SimpleCache
import com.google.android.exoplayer2.util.Util
import java.io.File
import java.util.*

class ExoMediaSourceHelper private constructor(context: Context) {
    private val mUserAgent: String
    private val mAppContext: Context
    private var mHttpDataSourceFactory: HttpDataSource.Factory? = null
    private var mCache: Cache? = null

    init {
        mAppContext = context.applicationContext
        mUserAgent = Util.getUserAgent(mAppContext, mAppContext.applicationInfo.name)
    }

    fun getMediaSource(uri: String): MediaSource {
        return getMediaSource(uri, null, false)
    }

    fun getMediaSource(uri: String, headers: MutableMap<String, String>?): MediaSource {
        return getMediaSource(uri, headers, false)
    }

    fun getMediaSource(uri: String, isCache: Boolean): MediaSource {
        return getMediaSource(uri, null, isCache)
    }

    fun getMediaSource(
        uri: String,
        headers: MutableMap<String, String>?,
        isCache: Boolean
    ): MediaSource {
        val contentUri = Uri.parse(uri)
        if ("rtmp" == contentUri.scheme) {
            return ProgressiveMediaSource.Factory(RtmpDataSource.Factory())
                .createMediaSource(MediaItem.fromUri(contentUri))
        } else if ("rtsp" == contentUri.scheme) {
            return RtspMediaSource.Factory().createMediaSource(MediaItem.fromUri(contentUri))
        }
        val contentType = inferContentType(uri)
        val factory: DataSource.Factory = if (isCache) {
            cacheDataSourceFactory
        } else {
            dataSourceFactory
        }
        if (mHttpDataSourceFactory != null) {
            setHeaders(headers)
        }
        return when (contentType) {
            C.CONTENT_TYPE_DASH -> DashMediaSource.Factory(factory)
                .createMediaSource(MediaItem.fromUri(contentUri))
            C.CONTENT_TYPE_HLS -> HlsMediaSource.Factory(factory)
                .createMediaSource(MediaItem.fromUri(contentUri))
            C.CONTENT_TYPE_OTHER -> ProgressiveMediaSource.Factory(factory)
                .createMediaSource(MediaItem.fromUri(contentUri))
            else -> ProgressiveMediaSource.Factory(factory)
                .createMediaSource(MediaItem.fromUri(contentUri))
        }
    }

    private fun inferContentType(fileName: String): Int {
        var fileName = fileName
        fileName = fileName.lowercase(Locale.getDefault())
        return if (fileName.contains(".mpd")) {
            C.CONTENT_TYPE_DASH
        } else if (fileName.contains(".m3u8")) {
            C.CONTENT_TYPE_HLS
        } else {
            C.CONTENT_TYPE_OTHER
        }
    }

    private val cacheDataSourceFactory: DataSource.Factory
        private get() {
            if (mCache == null) {
                mCache = newCache()
            }
            return CacheDataSource.Factory()
                .setCache(mCache!!)
                .setUpstreamDataSourceFactory(dataSourceFactory)
                .setFlags(CacheDataSource.FLAG_IGNORE_CACHE_ON_ERROR)
        }

    private fun newCache(): Cache {
        return SimpleCache(
            File(mAppContext.externalCacheDir, "exo-video-cache"),  //缓存目录
            LeastRecentlyUsedCacheEvictor(512 * 1024 * 1024),  //缓存大小，默认512M，使用LRU算法实现
            StandaloneDatabaseProvider(mAppContext)
        )
    }

    /**
     * Returns a new DataSource factory.
     *
     * @return A new DataSource factory.
     */
    private val dataSourceFactory: DataSource.Factory
        private get() = DefaultDataSource.Factory(mAppContext, httpDataSourceFactory)

    /**
     * Returns a new HttpDataSource factory.
     *
     * @return A new HttpDataSource factory.
     */
    private val httpDataSourceFactory: DataSource.Factory
        private get() {
            if (mHttpDataSourceFactory == null) {
                mHttpDataSourceFactory = DefaultHttpDataSource.Factory()
                    .setUserAgent(mUserAgent)
                    .setAllowCrossProtocolRedirects(true)
            }
            return mHttpDataSourceFactory!!
        }

    private fun setHeaders(headers: MutableMap<String, String>?) {
        if (headers != null && headers.size > 0) {
            //如果发现用户通过header传递了UA，则强行将HttpDataSourceFactory里面的userAgent字段替换成用户的
            if (headers.containsKey("User-Agent")) {
                val value = headers.remove("User-Agent")
                if (!TextUtils.isEmpty(value)) {
                    try {
                        val userAgentField =
                            mHttpDataSourceFactory!!.javaClass.getDeclaredField("userAgent")
                        userAgentField.isAccessible = true
                        userAgentField[mHttpDataSourceFactory] = value
                    } catch (e: Exception) {
                        //ignore
                    }
                }
            }
            mHttpDataSourceFactory!!.setDefaultRequestProperties(headers)
        }
    }

    fun setCache(cache: Cache?) {
        mCache = cache
    }

    companion object {
        @Volatile
        private var sInstance: ExoMediaSourceHelper? = null
        @JvmStatic
        fun getInstance(context: Context): ExoMediaSourceHelper {
            if (sInstance == null) {
                synchronized(ExoMediaSourceHelper::class.java) {
                    if (sInstance == null) {
                        sInstance = ExoMediaSourceHelper(context)
                    }
                }
            }
            return sInstance!!
        }
    }
}